lib/std/math/gamma.zig

```diff
@@ -0,0 +1,328 @@
+    if (abs < upper_bound) {
+        return @log(@abs(gamma(T, x)));
+    }
+
+    const log_base = @log(abs + lanczos_minus_half) - 1;
+    const exponent = abs - 0.5;
+    const log_series = @log(series(T, abs));
+    const initial = exponent * log_base + log_series - lanczos;
+
+    // use reflection formula for negatives
+    if (x < 0) {
+        const reflected = std.math.pi / (abs * sinpi(T, abs));
+        return @log(@abs(reflected)) - initial;
+    }
+    return initial;
+}
+
+// table of factorials for integer early return
+// stops at 22 because 23 isn't representable with full precision on f64
+const integer_result_table = [_]f64{
+    std.math.inf(f64), // gamma(+0.0)
+    1, // gamma(1)
+    1, // ...
+    2,
+    6,
+    24,
+    120,
+    720,
+    5040,
+    40320,
+    362880,
+    3628800,
+    39916800,
+    479001600,
+    6227020800,
+    87178291200,
+    1307674368000,
+    20922789888000,
+    355687428096000,
+    6402373705728000,
+    121645100408832000,
+    2432902008176640000,
+    51090942171709440000, // gamma(22)
+};
+
+// "g" constant, arbitrary
+const lanczos = 6.024680040776729583740234375;
+const lanczos_minus_half = lanczos - 0.5;
+
+fn series(comptime T: type, abs: T) T {
+    const numerator = [_]T{
+        23531376880.410759688572007674451636754734846804940,
+        42919803642.649098768957899047001988850926355848959,
+        35711959237.355668049440185451547166705960488635843,
+        17921034426.037209699919755754458931112671403265390,
+        6039542586.3520280050642916443072979210699388420708,
+        1439720407.3117216736632230727949123939715485786772,
+        248874557.86205415651146038641322942321632125127801,
+        31426415.585400194380614231628318205362874684987640,
+        2876370.6289353724412254090516208496135991145378768,
+        186056.26539522349504029498971604569928220784236328,
+        8071.6720023658162106380029022722506138218516325024,
+        210.82427775157934587250973392071336271166969580291,
+        2.5066282746310002701649081771338373386264310793408,
+    };
+    const denominator = [_]T{
+        0,
+        39916800,
+        120543840,
+        150917976,
+        105258076,
+        45995730,
+        13339535,
+        2637558,
+        357423,
+        32670,
+        1925,
+        66,
+        1,
+    };
+    var num: T = 0;
+    var den: T = 0;
+    // split to avoid overflow
+    if (abs < 8) {
+        // big abs would overflow here
+        for (0..numerator.len) |i| {
+            num = num * abs + numerator[numerator.len - 1 - i];
+            den = den * abs + denominator[numerator.len - 1 - i];
+        }
+    } else {
+        // small abs would overflow here
+        for (0..numerator.len) |i| {
+            num = num / abs + numerator[i];
+            den = den / abs + denominator[i];
+        }
+    }
+    return num / den;
+}
+
+// precise sin(pi * x)
+// but not for very small x and integer x, we handle those already
```

Comment:
how small is "very small" ?

Reply:
According to musl, it assumes `abs(x) > 2^-100`, but it only gets called for `abs(x) > 2^-54`, will add that to the comment